drivers: disk: stm32_sdmmc: add PM runtime support

Praneethsvch opened the pull request:
Added PM device runtime actions so driver now honors device runtime actions when PM_DEVICE and/or PM_DEVICE_RUNTIME is used. This allows dynamic suspend and resume actions. 

- `stm32_sdmmc_pm_action()` handling for PM_DEVICE_ACTION_RESUME/SUSPEND
- Runtime power management via pm_device_runtime_enable()
- Automatic clock gating and power GPIO control during suspend/resume
- Safe synchronization with DMA and IRQ state transitions
- Optional system PM policy locks around active transfers

This ensures the SDMMC peripheral enters low-power states when idle and
restores correctly when accessed again via disk or FS layers.

Tested on: STM32U575 (nucleo_u575zi_q)

etienne-lms commented on drivers/disk/sdmmc_stm32.c:
```diff
@@ -372,6 +374,9 @@ static int stm32_sdmmc_access_init(struct disk_info *disk)
 
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	pm_device_runtime_get(dev);
+#endif
```
`pm_device_runtime_*()` are already stubbed when `CONFIG_PM_DEVICE_RUNTIME` is disabled.
```suggestion
	pm_device_runtime_get(dev);
```
Ditto below.

etienne-lms commented on drivers/disk/sdmmc_stm32.c:
```diff
@@ -923,6 +933,163 @@ void stm32_sdmmc_get_card_csd(const struct device *dev, uint32_t csd[4])
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		err = stm32_sdmmc_activate(dev);
```
Nitpicking: can remove `err` and use `return stm32_sdmmc_xxx(dev);` straight.

etienne-lms commented on drivers/disk/sdmmc_stm32.c:
```diff
@@ -923,6 +933,163 @@ void stm32_sdmmc_get_card_csd(const struct device *dev, uint32_t csd[4])
 
+static int stm32_sdmmc_activate(const struct device *dev)
+{
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
```
Is this needed?

Mot of the sequence is already done in in `stm32_sdmmc_access_init()`. Could you factorize their common code?

bearsh reviewed (changes requested):
`pm_device_runtime_get()` / `pm_device_runtime_put()` is not called in `stm32_sdmmc_access_read()` `stm32_sdmmc_access_write()`

bearsh commented:
Hi @Praneethsvch 
it seems there a still issues with this PR... 

Personally, I would like to have as much PM related stuff merge as possible (and feasible) but I'm not a subsystem maintainer...

it would be nice to have statement from the st guys (maybe @erwango) how to proceed

Praneethsvch committed:
stm32_sdmmc: Add PM device runtime support

Support for PM device runtime is added.

Previously the driver behaviour is mostly always-on peripheral
after disk init. stm32_sdmmc_access_init() leaves the hardware on
until explicit deinit or card removal is performed. Meaning idle
current is high when disk is mounted as the clock tree, pin ctrl
are still active even when using PM device runtime.

Wih PM support, the driver gets an explicit hardware activate/deactivate
cycle through PM callbacks. Below are the improvements:
- Runtime PM enabling suspending the device when idle without powering
  down or calling deinit to save power.
- System PM can suspend/enable the device in a defined way.
- Refactor and improve pm locking.

Signed-off-by: Praneethsvch <praneeth_ch@icloud.com>

erwango commented:
Heads up, we plan to migrate SDMMC STM32 to SDHC driver (https://github.com/zephyrproject-rtos/zephyr/pull/101617). I hope it could be merged in coming weeks.
Target is to deprecate existing SDMMC driver as part of this release, so I'd advise to check this new driver and see if PM implementation could be adapted to this driver.
